### PR TITLE
Homework - Cars

### PR DESCRIPTION
**Pull request by LogarithmDerivative**
1. Added Engine class (which at the end I think is a bit overkill for this problem.
2. Fixed memory leaks by holding engines in unique_ptr.
3. Implemented getters for all of parameters.
4. Added new member speed_.
5. Brake function now sets speed_ to 0.
6. New exceptions along with proper validation:
- InvalidGear, 
- InvalidAcceleration, 
- InvalidEngineChange,
7. Encapsulated data.
8. Unit tests.
9. Workflow.

### CODE REVIEW

**Review by MaciejSalamonski — APPROVED** [review]
I dont have any comments ;) Nice work ~:)

File: Engine.hpp [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+class Engine {
+private:
+    int power_{};
+
+public:
+    Engine(int power);
+    int getPower() const { return power_; }
+    ~Engine() {};
+};
```
Comment (jakjan95) [inline Engine.hpp:11]: I am not sure about Engine class here, can generate problems and force us to have additional logic for Engine checking

File: PetrolEngine.cpp [hunk 2]
```diff
@@ -1,20 +1,18 @@
...
 
-void PetrolEngine::changeGear(int gear)
-{
-    // TODO: Add checking if gear is between -1 and gears_
-    // -1 is for REAR
-    // 0 is for NEUTRAL
+void PetrolEngine::changeGear(int gear) {
+    if (gear < -1 || gear > gears_) {
+        throw InvalidGear("This car does not have this gear.");
+    } else if (getCurrentGear() == gears_ && gear == -1) {
```
Comment (jakjan95) [inline PetrolEngine.cpp:13]: -1 is magic value

File: PetrolCar-ut.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+TEST_F(PetrolCarTest, shouldThrowExceptionWhenChangingToInvalidGear) {
+    ASSERT_THROW(car.getEngine()->changeGear(invalidValue), InvalidGear);
+    ASSERT_THROW(car.getEngine()->changeGear(-invalidValue), InvalidGear);
+}
+
+TEST_F(PetrolCarTest, shouldChangeGearAndThrowExceptionWhenChangingFrom5ToRear) {
+    car.getEngine()->changeGear(5);
+    ASSERT_EQ(car.getEngine()->getCurrentGear(), 5);
+    ASSERT_THROW(car.getEngine()->changeGear(-1), InvalidGear);
+}
```
Comment (jakjan95) [inline PetrolCar-ut.cpp:44]: magic values here

File: PetrolCar-ut.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+};
+
+TEST_F(PetrolCarTest, shouldAccelerate) {
+    car.accelerate(expectedSpeed);
+    ASSERT_EQ(car.getSpeed(), expectedSpeed);
+}
+
+TEST_F(PetrolCarTest, shouldBrake) {
+    car.brake();
+    ASSERT_EQ(car.getSpeed(), 0);
```
Comment (jakjan95) [inline PetrolCar-ut.cpp:28]: magic value

File: HybridCar-ut.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+TEST_F(HybridCarTest, shouldThrowExceptionWhenChangingToInvalidGear) {
+    ASSERT_THROW(car.getPetrolEngine()->changeGear(invalidValue), InvalidGear);
+    ASSERT_THROW(car.getPetrolEngine()->changeGear(-invalidValue), InvalidGear);
+}
+
+TEST_F(HybridCarTest, shouldChangeGearAndThrowExceptionWhenChangingFrom5ToRear) {
+    car.getPetrolEngine()->changeGear(5);
+    ASSERT_EQ(car.getPetrolEngine()->getCurrentGear(), 5);
+    ASSERT_THROW(car.getPetrolEngine()->changeGear(-1), InvalidGear);
+}
```
Comment (jakjan95) [inline HybridCar-ut.cpp:49]: magic values, try to create something like reverseGear, fifthGear

File: ElectricCar.cpp [hunk 6]
```diff
@@ -1,16 +1,28 @@
...
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+void ElectricCar::restore() {
+    charge();
+}
 
+void ElectricCar::changeEngine(std::unique_ptr<ElectricEngine> newEngine) {
+    if (getSpeed() == 0) {
+        engine_.swap(newEngine);
```
Comment (jakjan95) [inline ElectricCar.cpp:24]: also magic value

File: ElectricCar-ut.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+};
+
+TEST_F(ElectricCarTest, shouldAccelerate) {
+    car.accelerate(expectedSpeed);
+    ASSERT_EQ(car.getSpeed(), expectedSpeed);
+}
+
+TEST_F(ElectricCarTest, shouldBrake) {
+    car.brake();
+    ASSERT_EQ(car.getSpeed(), 0);
```
Comment (jakjan95) [inline ElectricCar-ut.cpp:27]: magic value

**Review by jakjan95 — COMMENTED** [review]
Some comments from me, mostly magic value to fix :)

**Review by WBigaj — COMMENTED** [review]
I have checked your code and only problems are magic values mentioned by @jakjan95

File: Engine.hpp (hunk 1 above)
Comment (LogarithmDerivative) [reply Engine.hpp:11]: Yeah this is a bit of overkill, but turned out there are not that many problems so I will left this

File: HybridCar-ut.cpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    car.brake();
+    car.changeEngine(std::move(newEngine));
+    ASSERT_NE(currEngine->getPower(), car.getElectricEngine()->getPower());
+    ASSERT_NE(currEngine->getBatteryCap(), car.getElectricEngine()->getBatteryCap());
+}
+
+TEST_F(HybridCarTest, shoudlThrowExceptionWhenChangingEngineWhileSpeedAbove0) {
+    car.accelerate(expectedSpeed);
+    ASSERT_THROW(car.changeEngine(std::make_unique<ElectricEngine>(power + modifier, batteryCap + modifier)), InvalidEngineChange);
+    ASSERT_THROW(car.changeEngine(std::make_unique<PetrolEngine>(power + modifier, capacity + modifier, gears + modifier)), InvalidEngineChange);
```
Comment (ziobron) [inline HybridCar-ut.cpp:69]: stylistic: Too long lines. I would move `make_unique` functions above this lines and pass unique_ptrs here.

File: PetrolCar-ut.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    ASSERT_THROW(car.accelerate(-expectedSpeed), InvalidAcceleration);
+}
+
+TEST_F(PetrolCarTest, shouldThrowExceptionWhenChangingToInvalidGear) {
+    ASSERT_THROW(car.getEngine()->changeGear(invalidValue), InvalidGear);
+    ASSERT_THROW(car.getEngine()->changeGear(-invalidValue), InvalidGear);
+}
+
+TEST_F(PetrolCarTest, shouldChangeGearAndThrowExceptionWhenChangingFrom5ToRear) {
+    car.getEngine()->changeGear(gears);
```
Comment (ziobron) [inline PetrolCar-ut.cpp:43]: change gear is a method of PetrolEngine, so in fact it should be in PetrolEngine unit tests.
Here you could have a delegation. If PetrolCar would have changeGear function it should just call changeGear on an Engine/Gearbox

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
7 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 1119, 1119, 1119] }

**Pull request closed without merging**